pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java

```diff
@@ -0,0 +1,364 @@
+    }
+    if (StringUtils.isNotBlank(authToken)) {
+      requestBuilder.addHeader(AUTH_HTTP_HEADER, authToken);
+    }
+    setTimeout(requestBuilder, DELETE_REQUEST_SOCKET_TIMEOUT_MS);
+    return sendRequest(requestBuilder.build());
+  }
+
+  // --------------------------------------------------------------------------
+  // JSON post/put utility APIs
+  // --------------------------------------------------------------------------
+
+  public SimpleHttpResponse postJsonRequest(URI uri, @Nullable String jsonRequestBody)
+      throws IOException {
+    return postJsonRequest(uri, jsonRequestBody, null);
+  }
+
+  public SimpleHttpResponse postJsonRequest(URI uri, @Nullable String requestBody,
+      @Nullable Map<String, String> headers)
+      throws IOException {
+    if (MapUtils.isEmpty(headers)) {
+      headers = new HashMap<>();
+    }
+    headers.put(HttpHeaders.CONTENT_TYPE, JSON_CONTENT_TYPE);
+    HttpEntity entity = requestBody == null ? null : new StringEntity(requestBody, ContentType.APPLICATION_JSON);
+    return sendPostRequest(uri, entity, headers, null);
+  }
+
+  public SimpleHttpResponse putJsonRequest(URI uri, @Nullable String jsonRequestBody)
+      throws IOException {
+    return putJsonRequest(uri, jsonRequestBody, null);
+  }
+
+  public SimpleHttpResponse putJsonRequest(URI uri, @Nullable String requestBody,
+      @Nullable Map<String, String> headers)
+      throws IOException {
+    if (MapUtils.isEmpty(headers)) {
+      headers = new HashMap<>();
+    }
+    headers.put(HttpHeaders.CONTENT_TYPE, JSON_CONTENT_TYPE);
+    HttpEntity entity = requestBody == null ? null : new StringEntity(requestBody, ContentType.APPLICATION_JSON);
+    return sendPutRequest(uri, entity, headers, null);
+  }
+
+  // --------------------------------------------------------------------------
+  // Lower-level request/execute APIs.
+  // --------------------------------------------------------------------------
+
+  public SimpleHttpResponse sendRequest(HttpUriRequest request)
+      throws IOException {
+    try (CloseableHttpResponse response = _httpClient.execute(request)) {
+      if (response.containsHeader(CommonConstants.Controller.HOST_HTTP_HEADER)) {
+        String controllerHost = response.getFirstHeader(CommonConstants.Controller.HOST_HTTP_HEADER).getValue();
+        String controllerVersion = response.getFirstHeader(CommonConstants.Controller.VERSION_HTTP_HEADER).getValue();
+        LOGGER.info(String
+            .format("Sending request: %s to controller: %s, version: %s", request.getURI(), controllerHost,
+                controllerVersion));
+      }
+      int statusCode = response.getStatusLine().getStatusCode();
+      if (statusCode >= 300) {
+        return new SimpleHttpResponse(statusCode, getErrorMessage(request, response));
+      }
+      return new SimpleHttpResponse(statusCode, EntityUtils.toString(response.getEntity()));
+    }
+  }
+
+  public CloseableHttpResponse execute(HttpUriRequest request)
```

Comment:
Seems this is only used in `FileUploadDownloadClient.downloadFile()`, I'd suggest moving that method here as well. We should try to avoid exposing this API as caller needs to close the response